integrations/pjrt/src/iree_pjrt/common/dylib_platform.cc

```diff
@@ -98,7 +99,15 @@ iree_status_t DylibPlatform::SubclassInitialize() {
     message.append(*loaded_compiler);
     logger().debug(message);
   }
-  compiler_ = std::make_unique<IREECompiler>();
+
+  std::vector<std::string> extra_compiler_options;
+  if (auto options_str = config_vars().Lookup("IREE_COMPILER_OPTIONS")) {
```

Comment:
Yeah, I've also considered this problem. Environment variables seems too global and difficult to control precisely, per-compilation.

And I found that, the main approach of configuring XLA in JAX (or other programs using XLA) is to set the environment variable `XLA_FLAGS`, documented in [List of XLA compiler flags](https://jax.readthedocs.io/en/latest/xla_flags.html#configuring-xla-in-jax). (e.g. [how alphafold3 use it](https://github.com/google-deepmind/alphafold3/blob/adb131f010d0b011586bc6caf3eda8e822e02f5a/docker/Dockerfile#L59))

However, these options can be overridden by [`debug_options`](https://github.com/openxla/xla/blob/30d8870c7cb2074899fea74adaa0656b6f0c9e1e/xla/pjrt/compile_options.proto#L32) (defined [here](https://github.com/openxla/xla/blob/30d8870c7cb2074899fea74adaa0656b6f0c9e1e/xla/xla.proto#L40)) and [`env_option_overrides`](https://github.com/openxla/xla/blob/30d8870c7cb2074899fea74adaa0656b6f0c9e1e/xla/pjrt/compile_options.proto#L150) (seems to mean overriding `XLA_FLAGS`). While `debug_options` is too XLA specific to use in IREE, `env_option_overrides` seems just a string map and can be used in IREE for per-compilation option configuring. (e.g. in JAX maybe we can pass it via `compiler_options` in [`jax.jit`](https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html), shown as an example below.)

```
@jax.jit(compiler_options={'iree-global-opt-propagate-transposes': True})
def f(x, y):
  ...
```

If we don't have an environment variable but always configure them per-compilation, there is a problem: most of these JAX programs assume that XLA as the backend compiler of JAX. Hence we need to change the code to add some IREE-specific compile flags. And if we added some IREE flags to the script, the code seems not "portable" (in aspect of the backend, i.e. XLA or IREE) anymore so we need to remove them to make it work in XLA again.

So in my view, maybe we can provide an environment variable while also allowing users to pass compile options via something like `env_option_overrides` so that on one hand users can add IREE-specific options without intrusive modifications, and on the other hand users can control compile options more percisely via parameters of functions like `jax.jit`.

Reply:
Maybe see how the other PJRT implementations (https://openxla.org/xla/pjrt/examples) handle this.

By the way, we're in a similar position with our PyTorch backend. PyTorch has both `torch.export` (AOT) and `torch.compile` (JIT). We've put the most mileage on torch.export since it gives us full control, while torch.compile has to conform to PyTorch semantics and work with existing programs, so we haven't invested as much in making it work well out of the box yet.

* https://pytorch.org/docs/stable/export.html
* https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html
* https://pytorch.org/docs/stable/torch.compiler_custom_backends.html

Depending on the use case, an AOT approach using `jax.export` (https://openxla.org/stablehlo/tutorials/jax-export) would sidestep some of these API questions.

Reply:
Yeah we can definitely export StableHLO from JAX and then control the compilation and execution percisely via IREE's own compiler/runtime APIs or tools, without PJRT. Maybe in the future we can provide two ways: use PJRT if users don't want to change any code -- just run the original code, but in IREE (with some trade-off: need to conform to restrictions/semantics of PJRT APIs); or use some IREE-specific wrappers for JAX, to export it first and then make some customization.

I checked some implementation: in XLA itself, it will read options from `XLA_FLAGS` as initial `debug_options` and users can pass `env_option_overrides` to override some values of `debug_options` (it's called debug options in XLA but is actually just compile options). In [Tenstorrent's impl](https://github.com/tenstorrent/tt-xla/blob/5f3200001213c476f8f17c92feb1b47ceca8a062/src/common/api_impl.cc#L608), seems compile options will be just ignored (if I didn't miss anything : ). I'm currently also reading some code from Intel and hope I can have some finding.

Since reading options from `env_option_overrides` seems a more elegant way rather than env vars, I think maybe I can open another PR to bring such feature first, and at the same time we can continue to discuss if such env var for compile options is suitable and if there's a better way : )

Reply:
> Maybe in the future we can provide two ways: use PJRT if users don't want to change any code -- just run the original code, but in IREE (with some trade-off: need to conform to restrictions/semantics of PJRT APIs); or use some IREE-specific wrappers for JAX, to export it first and then make some customization.

FYI: we had the https://github.com/iree-org/iree-jax project with "IREE-specific wrappers for JAX" at one point. That predated PJRT (and PyTorch 2.0, which added torch.compile and torch.export). It would be possible to revive that work (or redo it, given how long ago that was), if there is enough interest.

> Since reading options from `env_option_overrides` seems a more elegant way rather than env vars, I think maybe I can open another PR to bring such feature for review first, and at the same time we can continue to discuss if such env var for compile options is suitable and if there's a better way : )

I'm comfortable adding an environment variable backdoor, since that is useful for developers and tests. It's good to also have the front door for most users to use.